### PR TITLE
Expand CORINFO_HELP_CHKCASTANY cast

### PR DESCRIPTION
**Pull request by EgorBo**
I am curious if we can expand this helper too, e.g.:
```csharp
[Benchmark]
public void Test() => GenericCast<string>("");


[MethodImpl(MethodImplOptions.NoInlining)]
public T GenericCast<T>(object o) => (T)o;
```

Codegen diff for `GenericCast`. Basically, we use T as a fast-path check (with null check)

```diff
; Program.GenericCast[[System.__Canon, System.Private.CoreLib]](System.Object)
       sub       rsp,28
       mov       [rsp+20],rdx
+      mov       rax,r8
+      test      rax,rax
+      je        short M01_L00
       mov       rcx,[rdx+10]
       mov       rcx,[rcx]
+      cmp       [rax],rcx
+      je        short M01_L00
       mov       rdx,r8
       call      [CastHelpers.ChkCastAny(Void*, System.Object)]
       nop
       add       rsp,28
       ret
-; Total bytes of code 31
+; Total bytes of code 44
```

| Method |                   Toolchain |     Mean | Ratio | Code Size |
|------- |---------------------------- |---------:|------:|----------:|
|   Test |   \Core_Root_PR\corerun.exe | **1.120 ns** |  1.00 |      80 B |
|   Test | \Core_Root_base\corerun.exe | 1.500 ns |  1.34 |      67 B |

Basically, we inline these things:

image

(Technically, we can introduce a new helper without the fast-paths)

### CODE REVIEW

**Comment by EgorBo** [comment]
@VSadov @jkotas any concerns about it?

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -5486,7 +5486,8 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
         else if (isCastClass)
         {
             // Jit can only inline expand CHKCASTCLASS and CHKCASTARRAY helpers.
-            canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS) || (helper == CORINFO_HELP_CHKCASTARRAY);
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:5489]: This inlining can be profitable only when the type we are casting to is a class. It is not profitable when the type we are casting to is not a class. For example, it is not profitable when we are casting to `IEnumerable<Canon>` - it is a pure overhead in that case.

**Comment by jkotas** [comment]
It would be useful to have some evidence about how often the fast path is taken vs. not taken in something real world (e.g. Bing).

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -5486,7 +5486,8 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
...
+            canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS) || (helper == CORINFO_HELP_CHKCASTARRAY) ||
+                              (helper == CORINFO_HELP_CHKCASTANY);
         }
         else if ((helper == CORINFO_HELP_ISINSTANCEOFCLASS) || (helper == CORINFO_HELP_ISINSTANCEOFARRAY))
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:5492]: Should this be done for isinst too?

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -5625,7 +5626,7 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
     if (isCastClass)
     {
         assert((helper == CORINFO_HELP_CHKCASTCLASS) || (helper == CORINFO_HELP_CHKCASTARRAY) ||
-               (helper == CORINFO_HELP_CHKCASTINTERFACE));
+               (helper == CORINFO_HELP_CHKCASTANY) || (helper == CORINFO_HELP_CHKCASTINTERFACE));
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:5629]: Is sounds like a bug that we can end up here for `CORINFO_HELP_CHKCASTINTERFACE`. It does not make sense to inline the MT check for interface casts - it is always going to be false.

Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5629]: It is used for profiled casts, e.g.:

```csharp
static IDisposable CastInterface(object o)
{
    return (IDisposable)o;
}
```
emits:
```asm
       sub      rsp, 40
       mov      rax, rcx
       test     rax, rax
       je       SHORT G_M63531_IG05
       mov      rdx, 0x7FFBDF89AE08      ; MyDisposable (class implementing IDisposable)
       cmp      qword ptr [rax], rdx
       je       SHORT G_M63531_IG05
       mov      rdx, rcx
       mov      rcx, 0x7FFBDEFF1C30      ; System.IDisposable
       call     [CORINFO_HELP_CHKCASTINTERFACE]
G_M63531_IG05:  
       nop      
       add      rsp, 40
       ret      
; Total bytes of code 52
```
Cast profiling is only enabled for static pgo training (`DOTNET_JitProfileCasts=1`)

File: src/coreclr/jit/importer.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5489]: Presumably the overhead is small or can be eliminated completely if we introduce a slightly stripped version of fallback?

I'll check the stats which path is more often taken in a large service like you requested

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5489]: My primary point was that the inlining of the exact check is not profitable when you can statically tell at JIT time that the target is not a class.

You are right that we can go even further and introduce specialized helper that omits the exact class check. I am not sure whether it is worth it given the overall cost of the helper.

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5489]: For example, we should not be inlining the exact check in:
```
[MethodImpl(MethodImplOptions.NoInlining)]
public IEnumrable<T> GenericCast<T>(object o) => (IEnumerable<T>)o;
```

**Comment by EgorBo** [comment]
> It would be useful to have some evidence about how often the fast path is taken vs. not taken in something real world (e.g. Bing).

@jkotas 

I've reverted this PR's change and put this lame counters into the ChkCastAny:

image

Here is the **Bing**'s stats (after running its benchmark):
```
total=220170000, isNull=117642, isSame=39514246;
```
so in 18% cases the fast path is taken (0.05% for null)

**AvaloniaILSpy** desktop app:
```
total=75000, isNull=14237, isSame=28136;
```
36% for the fast path.

**Comment by jkotas** [comment]
Sounds good. These percentages should be even higher once you do the expansion only in places where it is statically known to be profitable.

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5492]: Probably? The Bing's stats for `IsInstanceOfAny` are:
```
ttotal=25578135, iisNull=2278487, iisSame=3925180

15% same
10% is null
```

Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5499]: I didn't add `IsInstanceOfAny` because we don't oportunistically expand it even for non-generics today, e.g.
```csharp
bool isinst = obj is MyClass1;
```
(we only do that when `MyClass1` is sealed and won't even emit a fallback)

we only do that for castclass, e.g. `(MyClass1)obj`. So presumably I'll have to slightly alter the logic to support `IsInstanceOfAny` + fallback

**Review by jkotas — APPROVED** [review]
